pkg/config/config.go

```diff
@@ -0,0 +1,303 @@
+	K8s: K8sConfig{
+		Context:   "",
+		Namespace: "",
+	},
+	Worker: WorkerConfig{
+		Steps: StepsConfig{
+			Docker: DockerStepConfig{
+				Image: "gcr.io/kaniko-project/executor:v1.7.0",
+			},
+			Kubectl: KubectlStepConfig{
+				Image: "docker.io/wharfse/kubectl:v1.23.5",
+			},
+			Helm: HelmStepConfig{
+				Image: "docker.io/wharfse/helm:v3.8.1",
+			},
+		},
+	},
+	Provisioner: ProvisionerConfig{
+		HTTP: HTTPConfig{
+			CORS: CORSConfig{
+				AllowAllOrigins: false,
+				AllowOrigins:    []string{},
+			},
+			BindAddress: "0.0.0.0:5009",
+		},
+		Worker: ProvisionerWorkerConfig{
+			ServiceAccountName: "wharf-cmd",
+			InitContainer: K8sContainerConfig{
+				Image:           "bitnami/git",
+				ImageTag:        "2-debian-10",
+				ImagePullPolicy: v1.PullIfNotPresent,
+			},
+			Container: K8sContainerConfig{
+				Image:           "quay.io/iver-wharf/wharf-cmd",
+				ImageTag:        "latest",
+				ImagePullPolicy: v1.PullAlways,
+			},
+		},
+	},
+	Aggregator: AggregatorConfig{
+		WharfAPIURL:           "http://wharf-api:8080",
+		WorkerAPIExternalPort: "5010",
+	},
+	Watchdog: WatchdogConfig{
+		WharfAPIURL:    "http://wharf-api:8080",
+		ProvisionerURL: "http://wharf-cmd-provisioner:8080",
+	},
+}
+
+// LoadConfig looks for, parses and validates the config and returns it as a
+// Config object.
+func LoadConfig() (Config, error) {
+	cfgBuilder := config.NewBuilder(DefaultConfig)
+	// TODO: Also add config file relative to home, e.g.:
+	//  ~/.config/iver-wharf/wharf-cmd/wharf-cmd-config.yml
+	//  $HOME/.config/iver-wharf/wharf-cmd/wharf-cmd-config.yml
+	// And OS-specific config paths.
+	cfgBuilder.AddConfigYAMLFile("/etc/iver-wharf/wharf-cmd/wharf-cmd-config.yml")
+	cfgBuilder.AddConfigYAMLFile(".wharf-cmd-config.yml")
+	if cfgFile, ok := os.LookupEnv("WHARF_CONFIG"); ok {
+		cfgBuilder.AddConfigYAMLFile(cfgFile)
+	}
+	cfgBuilder.AddEnvironmentVariables("WHARF")
+
+	var cfg Config
+	if err := cfgBuilder.Unmarshal(&cfg); err != nil {
+		return Config{}, fmt.Errorf("load config: %w", err)
+	}
+	if err := cfg.validate(); err != nil {
+		return Config{}, fmt.Errorf("load config: %w", err)
+	}
+
+	return cfg, nil
+}
+
+func (c Config) validate() error {
+	initContainerPullPolicy := c.Provisioner.Worker.InitContainer.ImagePullPolicy
+	if !parseImagePolicy(&initContainerPullPolicy) {
+		return fmt.Errorf("invalid pull policy: provisioner.worker.initContainer.imagePullPolicy=%s", initContainerPullPolicy)
+	}
+	containerPullPolicy := c.Provisioner.Worker.Container.ImagePullPolicy
+	if !parseImagePolicy(&containerPullPolicy) {
+		return fmt.Errorf("invalid pull policy: provisioner.worker.container.imagePullPolicy=%s", containerPullPolicy)
+	}
+	return nil
+}
+
+func parseImagePolicy(p *v1.PullPolicy) bool {
+	switch strings.ToLower(string(*p)) {
+	case "always":
+		*p = v1.PullAlways
+	case "never":
+		*p = v1.PullNever
+	case "ifnotpresent":
+		*p = v1.PullIfNotPresent
+	default:
+		return false
+	}
+	return true
+}
```

Comment:
Inside `parseImagePolicy` you're updating the value by pointer, but inside `Config.validate` you're passing in a reference to the local field, and the `Config.validate` method receiver isn't a pointer as well.

This "case insensitive parsing" doesn't actually update the values.

Kubernetes is still case insensitive, but our switch statements later down in the code might not be, if we would ever switch on them.

Suggest returning `(v1.PullPolicy, bool)` instead of doing pointer changes, and change `func (Config) validate()` to `func (*Config) validate()`.

I'm not a fan of changing values via pointers, as it's easily confused and overlooked. It's not that complicated of a technique, but the ease of mistakes outweigh the pros, IMO.

Reply:
Suggest also writing a unit test for `Config.validate`